### PR TITLE
Feat: #3498 Display episode's published_at

### PR DESCRIPTION
**Pull request by vanduc1102**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x ] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description
Display episode's published at 

## Related Tickets & Documents

https://github.com/thepracticaldev/dev.to/issues/3498

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

older-than-1-day
Screenshot at Oct 07 19-19-09
no-published-at

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [ x] no documentation needed

### CODE REVIEW

File: app/models/podcast_episode.rb [hunk 1]
```diff
@@ -86,6 +86,14 @@ def published
...
   end
 
+  def readable_publish_date
+    published_at.strftime("%b %e")
+  end
+
+  def published_timestamp
+    published_at.utc.iso8601
+  end
+
```
Comment (rhymes) [inline app/models/podcast_episode.rb]: These should go into the `podcast_episode_decorator.rb` file

Comment (vanduc1102) [reply app/models/podcast_episode.rb]: thanks @rhymes for helps, i'm working on the issue.

**Comment by vanduc1102** [comment]
> Thanks for the pr, @vanduc1102
> Regarding the failing tests, we are not actually attempting to display unpublished episodes in this case. All episodes are considered published if the podcast itself is published. The tests are failing because having `published_at` is not a strict requirement for a podcast episode record. So, before showing the date we should check if it exists. I also recommend adding a check to the new methods so that they wouldnt fail if we run them on a record that doesn't have `published_at`.
> It would be nice to add a test to check that the date is displayed correctly (to `spec/system/podcasts/*`.
> I will require changes though the pr is WIP to mark it for me and other reviewers.

Thanks @lightalloy, 
I made some changes
- Added check for published_at
- Added testcases
Would you please help to review again?

File: app/models/podcast_episode.rb (hunk 1 above)
Comment (vanduc1102) [reply app/models/podcast_episode.rb]: Done! thanks @rhymes

**Comment by vanduc1102** [comment]
> Hi @vanduc1102, thanks for this PR!
> 
> The tests are failing because they are attempting to render podcast episodes that are not published, which looks like a symptom of something else, because an unpublished episode shouldn't end up in the index page. The controller is actually loading only `available` episodes:
> 
> https://github.com/thepracticaldev/dev.to/blob/580fd11bb9b0e5e6fac0b3691989df3078c77a9a/app/controllers/podcast_episodes_controller.rb#L9-L11

Thanks @rhymes for taking your time to review my PR, 
it is my first PR in RnR, so it kinds of messy. 😭 
please help to review again.

File: app/decorators/podcast_episode_decorator.rb [hunk 2]
```diff
@@ -8,4 +8,12 @@ def comments_to_show_count
   def cached_tag_list_array
     (tag_list || "").split(", ")
   end
+
+  def readable_publish_date
+    published_at&.strftime("%b %e")
```
Comment (rhymes) [inline app/decorators/podcast_episode_decorator.rb]: I think this should display the year for older podcasts, see how for example, it is done for the comment:

```ruby
  def readable_publish_date
    if created_at.year == Time.current.year
      created_at.strftime("%b %e")
    else
      created_at.strftime("%b %e '%y")
    end
  end
```

File: app/views/articles/_podcast_episodes_feed.html.erb [hunk 3]
```diff
@@ -18,6 +18,12 @@
...
         <%= episode.podcast.title %>
+        <% if episode.published_at.present? %>
+          ・<time class="published-at" datetime="<%= episode.decorator.published_timestamp %>">
+          <%= episode.decorator.readable_publish_date %>
+          </time>
+          <span class="time-ago-indicator-initial-placeholder" data-seconds="<%= episode.published_at_int %>"></span>
+        <% end %>
       </h4>
     </div>
   </a>
```
Comment (rhymes) [inline app/views/articles/_podcast_episodes_feed.html.erb]: This file doesn't seem to be in use, you can remove it.

File: app/views/podcast_episodes/_episodes_feed.html.erb [hunk 4]
```diff
@@ -19,6 +19,12 @@
       <h4>
         <a href="/<%= episode.podcast.slug %>">
           <%= episode.podcast.title %>
+          <% if episode.published_at.present? %>
+            ・<time class="published-at" datetime="<%= episode.decorate.published_timestamp %>">
+            <%= episode.decorate.readable_publish_date %>
```
Comment (rhymes) [inline app/views/podcast_episodes/_episodes_feed.html.erb]: you don't need to call `.decorate`, the collection has been already decorated in the controller

File: app/views/articles/_podcast_episodes_feed.html.erb [hunk 5]
```diff
@@ -18,6 +18,12 @@
       </div>
       <h4>
         <%= episode.podcast.title %>
+        <% if episode.published_at.present? %>
+          ・<time class="published-at" datetime="<%= episode.decorator.published_timestamp %>">
```
Comment (lightalloy) [inline app/views/articles/_podcast_episodes_feed.html.erb]: `decorator` doesn't seem like a valid method, so rendering this template should fail.
I checked the codebase and found out that it's not used in the codebase. I suppose we should delete it. Please, tell me if you find that it's used and I haven't noticed it.

File: app/views/podcast_episodes/_episodes_feed.html.erb (hunk 4 above)
Comment (lightalloy) [reply app/views/podcast_episodes/_episodes_feed.html.erb]: You will also need to decorate the collection inside the `StoriesController#handle_podcast_index`

File: app/views/podcast_episodes/show.html.erb [hunk 6]
```diff
@@ -50,6 +50,11 @@
       <% else %>
         <h1><%= @episode.title %></h1>
       <% end %>
+      <% if @episode.published_at.present? %>
```
Comment (lightalloy) [inline app/views/podcast_episodes/show.html.erb]: Btw, you can check it a bit shorter `@episode.published_at?`

File: app/views/podcast_episodes/show.html.erb [hunk 7]
```diff
@@ -50,6 +50,11 @@
...
+        <time class="published-at" datetime="<%= @episode.decorate.published_timestamp %>">
```
Comment (lightalloy) [inline app/views/podcast_episodes/show.html.erb]: I think it's worth decorating in the controller action (`StoriesController#handle_podcast_show`) and not decorate multiple times inside a template.

File: app/views/podcast_episodes/_episodes_feed.html.erb (hunk 4 above)
Comment (vanduc1102) [reply app/views/podcast_episodes/_episodes_feed.html.erb]: thanks, i'm working on it

**Comment by vanduc1102** [comment]
Hi @lightalloy @rhymes 
I addressed the comments

Screen Shot 2019-10-08 at 5 58 09 PM

File: app/views/articles/_podcast_episodes_feed.html.erb [hunk 8]
```diff
@@ -18,12 +18,6 @@
       </div>
```
Comment (rhymes) [inline app/views/articles/_podcast_episodes_feed.html.erb]: I meant to actually remove the file :D I can't find a place in the codebase where this is used

**Comment by rhymes** [comment]
@vanduc1102 thanks for the fix and the screenshot. I noticed something weird in the picture, the time ago is displayed only for some episodes in the list

File: app/controllers/stories_controller.rb [hunk 9]
```diff
@@ -141,7 +141,8 @@ def handle_podcast_index
     @podcast_index = true
     @article_index = true
     @list_of = "podcast-episodes"
-    @podcast_episodes = @podcast.podcast_episodes.reachable.order("published_at DESC").limit(30)
+    @podcast_episodes = PodcastEpisodeDecorator.decorate_collection(@podcast.podcast_episodes.
```
Comment (lightalloy) [inline app/controllers/stories_controller.rb]: I believe you can do that with just adding `.decorate` to the query. Haven't noticed that on the first review.

File: app/views/articles/_podcast_episodes_feed.html.erb (hunk 8 above)
Comment (lightalloy) [reply app/views/articles/_podcast_episodes_feed.html.erb]: Yeah, the file itself can be deleted :)

**Comment by vanduc1102** [comment]
> @vanduc1102 thanks for the fix and the screenshot. I noticed something weird in the picture, the time ago is displayed only for some episodes in the list

Hi @rhymes , i deleted the file `app/views/articles/_podcast_episodes_feed.html.erb`
`the time ago is displayed only for some episodes in the list`, i followed the implementation in https://github.com/thepracticaldev/dev.to/blob/master/app/views/articles/index.html.erb#L121 
which is rendering on dev.to live 
Screen Shot 2019-10-08 at 7 37 10 PM

File: app/controllers/stories_controller.rb (hunk 9 above)
Comment (vanduc1102) [reply app/controllers/stories_controller.rb]: would you give me example. i'm new to ruby

Comment (lightalloy) [reply app/controllers/stories_controller.rb]: I mean `@podcast.podcast_episodes.reachable.order("published_at DESC").limit(30).decorate`

File: app/views/articles/_podcast_episodes_feed.html.erb (hunk 8 above)
Comment (vanduc1102) [reply app/views/articles/_podcast_episodes_feed.html.erb]: thanks , deleted

File: app/controllers/stories_controller.rb (hunk 9 above)
Comment (vanduc1102) [reply app/controllers/stories_controller.rb]: Updated, thanks

**Comment by rhymes** [comment]
@vanduc1102 yeah you did right, i'll investigate because that sounds like a bug in both places. But it's out of the scope of this PR anyway!

**Pull request merged**